### PR TITLE
FIX: Avoid creating unnecessary `index` column when executing Aggregation node on Pandas backend

### PR DESCRIPTION
**Pull request by timothydijamco**
This is a really simple change, but a detailed explanation is unfortunately pretty long because of some nuances.

**TL;DR:** The Pandas backend sometimes does an unnecessary `reset_index` when executing the `Aggregation` node. This PR makes it only do `reset_index` when it actually should. This is simply an internal-logic correction that doesn't change any user-facing behavior.

# Detailed explanation
## Background

The Pandas backend's execution method for the `Aggregation` node is called `execute_aggregation_dataframe`.

`execute_aggregation_dataframe` does this:

1. [Executes the aggregation metrics requested by the user](https://github.com/ibis-project/ibis/blob/13387c5a0969c99be403af5b3e8aa70bdb671623/ibis/pandas/execution/generic.py#L426)
2. [Concats the metric results (`pd.Series`) into one `pd.DataFrame`](https://github.com/ibis-project/ibis/blob/13387c5a0969c99be403af5b3e8aa70bdb671623/ibis/pandas/execution/generic.py#L432)
3. [Calls `reset_index`](https://github.com/ibis-project/ibis/blob/13387c5a0969c99be403af5b3e8aa70bdb671623/ibis/pandas/execution/generic.py#L432)

The `reset_index` is great when **aggregating over groups,** because `reset_index` will make the resulting dataframe have a column containing the group-by keys (rather than having them in the index).

However, the `reset_index` is not as great when **aggregating with no grouping**, because `reset_index` will then create an unnecessary column named `index` (with values `0, 1, 2, ...`).

## This PR

This PR corrects `execute_aggregation_dataframe` so that it does not `reset_index` when aggregating with no grouping.

This avoids creating an unnecessary `index` column in the resulting dataframe (when aggregating with no grouping)

#### Note

This change is internal-facing.

[`ibis.pandas.core.execute`](https://github.com/ibis-project/ibis/blob/master/ibis/pandas/core.py#L349) will now output dataframes with no unnecessary `index` column.

However, [`ibis.pandas.client.PandasClient.execute`](https://github.com/ibis-project/ibis/blob/13387c5a0969c99be403af5b3e8aa70bdb671623/ibis/pandas/client.py#L359) (which most users would use) will have no change in behavior, since it uses [`ibis.pandas.core.execute_and_reset`](https://github.com/ibis-project/ibis/blob/13387c5a0969c99be403af5b3e8aa70bdb671623/ibis/pandas/core.py#L418), which is implemented in a way that _happens to_ get rid of columns that are not in the schema before returning the dataframe to user.

So why this change? This makes execution of `Aggregation` correct in and of itself. If using `ibis.pandas.core.execute` directly, a correct result will be obtained.

## Testing

I couldn't find comprehensive tests for `aggregate`, but I created some here. These tests are from the user's perspective, so they all pass both before and after this PR (again, no user-facing change in this PR). These new tests are just to help demonstrate this doesn't break anything.

### CODE REVIEW

**Comment by timothydijamco** [comment]
cc @icexelloss

File: ibis/pandas/execution/generic.py [hunk 1]
```diff
@@ -428,8 +428,12 @@ def execute_aggregation_dataframe(op, data, scope=None, **kwargs):
         for metric in op.metrics
     ]
 
-    # group by always needs a reset to get the grouping key back as a column
-    result = pd.concat(pieces, axis=1).reset_index()
+    result = pd.concat(pieces, axis=1)
+
+    # If grouping, need a reset to get the grouping key back as a column
+    if op.by:
```
Comment (icexelloss) [inline ibis/pandas/execution/generic.py:434]: Can you explain why this change have no user facing effect? Before this PR, is the column "index" being dropped sometime later?

File: ibis/tests/all/test_aggregation.py [hunk 2]
```diff
@@ -1,9 +1,73 @@
 import numpy as np
+import pandas as pd
 import pytest
 from pytest import param
 
 from ibis.tests.backends import Clickhouse, MySQL, Postgres, PySpark, SQLite
 
+aggregate_test_params = [
```
Comment (icexelloss) [inline ibis/tests/all/test_aggregation.py:8]: Are the tests here different from what's defined here?

https://github.com/ibis-project/ibis/blob/master/ibis/tests/all/test_aggregation.py#L132

Comment (timothydijamco) [reply ibis/tests/all/test_aggregation.py:8]: They are a bit different

`test_aggregation` tests API methods that create `Reduction` nodes/ops (e.g. [`mean`](https://github.com/ibis-project/ibis/blob/13387c5a0969c99be403af5b3e8aa70bdb671623/ibis/expr/operations.py#L873), [`min`](https://github.com/ibis-project/ibis/blob/13387c5a0969c99be403af5b3e8aa70bdb671623/ibis/expr/operations.py#L960), [`max`](https://github.com/ibis-project/ibis/blob/13387c5a0969c99be403af5b3e8aa70bdb671623/ibis/expr/operations.py#L954)). Execution of these uses `aggcontext.agg` ([example](https://github.com/ibis-project/ibis/blob/master/ibis/pandas/execution/generic.py#L460)) and returns a scalar

However, the new `test_aggregate` and `test_aggregate_grouped` in this PR test the `aggregate` API method, which creates an `Aggregation` node/op. This has its own execution method and returns a dataframe

Comment (timothydijamco) [reply ibis/tests/all/test_aggregation.py:8]: Let me rename `test_aggregation` to `test_reduction_ops`. I think that will be clearer.

File: ibis/pandas/execution/generic.py (hunk 1 above)
Comment (timothydijamco) [reply ibis/pandas/execution/generic.py:434]: Yep, the Ibis Pandas client uses [`ibis.pandas.core.execute_and_reset`](https://github.com/ibis-project/ibis/blob/13387c5a0969c99be403af5b3e8aa70bdb671623/ibis/pandas/core.py#L418), which gets rid of columns that are not in the schema.

(I don't think that function is directly intended to correct the output of node execution methods that return unnecessary columns. It's main goal is to reset the index of the final dataframe before returning it to the user, keeping it as a column if its in the schema, and dropping it otherwise. This just happens to fix the output of `execute_aggregation_dataframe`.)

File: ibis/tests/all/test_aggregation.py (hunk 2 above)
Comment (timothydijamco) [reply ibis/tests/all/test_aggregation.py:8]: Rename done

**Comment by timothydijamco** [comment]
Some changes to fix failing tests:
1. Remove use of `std` when testing `aggregate`. `std` doesn't work well with `test_aggregate_grouped` because the standard deviation is actually always 0 in each group (and some backends will have values closer/farther from 0 than Pandas calculates it)
2. Change `.sort_values(..., ignore_index=True)` to `.sort_values(...).reset_index(drop=True)` for Pandas <1.0.0 compatibility

File: ibis/pandas/execution/generic.py (hunk 1 above)
Comment (icexelloss) [reply ibis/pandas/execution/generic.py:434]: Yeah I think it's cleaner if the output of `execute_aggregation_dataframe` doesn't have a column called "index"

File: ibis/tests/all/test_aggregation.py (hunk 2 above)
Comment (icexelloss) [reply ibis/tests/all/test_aggregation.py:8]: I think what you have a pretty similar to;
https://github.com/ibis-project/ibis/blob/ec37e6e5ed14f4e7bc72dc2b70b07d85368f4376/ibis/tests/all/test_aggregation.py#L39

Can you try moving the new test to that file?

Comment (jreback) [reply ibis/tests/all/test_aggregation.py:8]: i agree, happy to have you split out this to a separate file (moving all grouping ops tests)

Comment (timothydijamco) [reply ibis/tests/all/test_aggregation.py:8]: @icexelloss Sorry I am a bit confused, would you mind clarifying? The new tests are actually already in `test_aggregation.py`. Possibly there's a mix-up from above where I mentioned renaming `test_aggregation` to `test_reduction_ops`—that actually was to rename a specific test that was named `test_aggregation` to `test_reduction_ops` (not renaming the file named `test_aggregation.py`). Are you OK with splitting the file up like @jreback is suggesting?

This is what makes up [`test_aggregation.py`](https://github.com/ibis-project/ibis/blob/ec37e6e5ed14f4e7bc72dc2b70b07d85368f4376/ibis/tests/all/test_aggregation.py):

- Tests for `Aggregation` op (ungrouped) — new in this CR
- Tests for `Aggregation` op (grouped) — new in this CR
- Tests for `Reduction` subclass ops
- Tests for `TopK` op (I think `TopK` doesn't quite fit into either of the above. It's an aggregation/reduction that returns `K` number of rows, but isn't an `Aggregation` op nor a `Reduction` op)

I could split the file up this way:
1. The `Aggregation` op tests and `TopK` op tests in a file called `test_aggregation.py`
2. The `Reduction` op tests in a file called `test_reduction.py`

Comment (icexelloss) [reply ibis/tests/all/test_aggregation.py:8]: Oh sorry I misread. Yes what you are doing now is what I suggested. So LGTM.

Comment (jreback) [reply ibis/tests/all/test_aggregation.py:8]: lol, i see now as well.
